### ISSUE TITLE
[FR] Add batch_ndims to numpyro.infer.util.log_density

### ISSUE THREAD

**Issue by amifalk** [pyro-ppl/numpyro#1674]
numpyro.infer.log_likelihood takes a batch dim argument that allows evaluating samples in parallel (or at least as a vectorized operation). It would be nice to have this functionality for log_density as well, for example, for finding the MAP after running MCMC or evaluating the posterior density of held-out samples.

I don't think it would be too hard to change the api to be match log_likelihood:
```python
log_density(model, posterior_samples, *args, parallel=False, batch_ndims=1, **kwargs)
```

If there's interest, I would be happy to take a stab at a PR.

**Comment by fehiepsi** [comment]
You can use vmap to achieve this purpose, so I guess this feature is unnecessary.

**Comment by amifalk** [comment]
Unfortunately vmapping the existing log_density function requires making a model-dependent pytree to tell jax to batch over all params, so it's not very ergonomic. I've hacked together a general solution using code from the log_likelihood function. The function is below in case anyone runs into the same issue.  

```python
import jax.numpy as jnp
from numpyro.handlers import substitute, trace
from jax.lax import broadcast_shapes
from numpyro.distributions.util import is_identically_one
from numpyro.infer.util import soft_vmap

def log_dens(model, posterior_samples, *args, parallel=False, batch_ndims=1, **kwargs):
    def single_log_density(samples):
        substituted_model = (
            substitute(model, samples) if isinstance(samples, dict) else model
        )
        model_trace = trace(substituted_model).get_trace(*args, **kwargs)
        
        log_joint = jnp.zeros(())
        
        for site in model_trace.values():
            if site["type"] == "sample":
                value = site["value"]
                intermediates = site["intermediates"]
                scale = site["scale"]
                if intermediates:
                    log_prob = site["fn"].log_prob(value, intermediates)
                else:
                    guide_shape = jnp.shape(value)
                    model_shape = tuple(
                        site["fn"].shape()
                    )  # TensorShape from tfp needs casting to tuple
                    try:
                        broadcast_shapes(guide_shape, model_shape)
                    except ValueError:
                        raise ValueError(
                            "Model and guide shapes disagree at site: '{}': {} vs {}".format(
                                site["name"], model_shape, guide_shape
                            )
                        )
                    log_prob = site["fn"].log_prob(value)

                if (scale is not None) and (not is_identically_one(scale)):
                    log_prob = scale * log_prob

                log_prob = jnp.sum(log_prob)
                log_joint = log_joint + log_prob
        
        return log_joint

    prototype_site = batch_shape = None
    for name, sample in posterior_samples.items():
        if batch_shape is not None and jnp.shape(sample)[:batch_ndims] != batch_shape:
            raise ValueError(
                f"Batch shapes at site {name} and {prototype_site} "
                f"should be the same, but got "
                f"{sample.shape[:batch_ndims]} and {batch_shape}"
            )
        else:
            prototype_site = name
            batch_shape = jnp.shape(sample)[:batch_ndims]    

    if batch_shape is None:  # posterior_samples is an empty dict
        batch_shape = (1,) * batch_ndims
        posterior_samples = np.zeros(batch_shape)

    batch_size = int(np.prod(batch_shape))
    chunk_size = batch_size if parallel else 1

    return soft_vmap(single_log_density, posterior_samples, len(batch_shape), chunk_size)
```

**Comment by fehiepsi** [comment]
I see, thanks for the code! We might need to use `lambda : ...` to avoid pytree semantics. The current `log_density` api also supports parameters, which is not batched, so I don't think we can change its api.
